### ISSUE TITLE
Optimize code for refreshing Board

### ISSUE THREAD

**Issue by tu2-atmanand** [tu2-atmanand/Task-Board#30]
Currently, i have to functions `refreshBoardData` and `refreshTasks`. I think its possible to improve this functions to refresh the Task Board complete view efficiently after i do any operation.
I was thinking of having two functions, one is to refresh the whole board and another to just refresh the tasks inside the currently active Board.
So basically the first function will first read the data.json to render the empty columns and then it will read the tasks.json to render all the tasks inside this columns based on criteria defined and also sorting them after implementing the feature.
But the second function will only read the tasks.json and render all the tasks inside the currently active board.

But the problem happened at present is i think, after the refreshBoardData function is executed the second function doesnt run after that, i think thats why i dont see the tasks updated inside the board. Also, right now i am using the method of *React.Dispatch* to set the boards and tasks. So i dont know if it is possible to call the above functions i have created from a file in another folder. In the branch opti-refreshingBoard, i am trying to implement this features.

**Comment by tu2-atmanand** [comment]
I have found the following solution for now to start with to see if this will resolve the issue : 

Your thinking is on the right track. To ensure that `refreshTasks` is executed after `refreshBoardData`, the task refresh logic should be triggered once the board data is successfully loaded. Since the `refreshBoardData` function is already being called when you press the `RefreshBtn`, you just need to ensure that it also triggers `refreshTasks` after the board data is set.

Here’s how you can modify the flow to optimize and ensure that `refreshTasks` is executed after `refreshBoardData`:

1. **Update `refreshBoardData`**:
   - Pass a callback function to `refreshBoardData` to call `refreshTasks` after the boards are loaded.

2. **Call `refreshTasks` after board refresh**:
   - Once the boards are refreshed, trigger `refreshTasks` for the relevant columns inside the `Column` component.

### Changes:

1. **In `refreshBoardData`** (add a callback to trigger `refreshTasks`):
   ```ts
   export const refreshBoardData = async (
       setBoards: React.Dispatch<React.SetStateAction<Board[]>>,
       callback: () => void // Add this callback
   ) => {
       console.log("------ Inside the refreshBoardData function -----");
       try {
           const data = await loadBoardsData(); // Fetch updated board data
           setBoards(data); // Update the state with the new data
           callback(); // Call the callback after boards are set
       } catch (err) {
           console.error("Failed to refresh boards data:", err);
       }
   };
   ```

2. **In `KanbanBoard.tsx`**:
   Update the `RefreshBtn` to ensure `refreshTasks` is called:
   ```tsx
   <button className="RefreshBtn" onClick={() => refreshBoardData(setBoards, () => {
       // Trigger refreshTasks after the boards are refreshed
       boards.forEach((board, index) => {
           board.columns.forEach((column) => {
               refreshTasks(setTasks, index, column.colType, column.data);
           });
       });
   })}>
       <RefreshCcw size={20} />
   </button>
   ```

3. **In `Column.tsx`**:
   You can keep the same `useEffect` to load tasks when the column is first rendered:
   ```tsx
   useEffect(() => {
       refreshTasks(setTasks, activeBoard, colType, data); // This will still run on initial load
   }, [colType, data]);
   ```

### Key Points:
- The `refreshBoardData` function now accepts a callback and calls it after setting the boards, ensuring that `refreshTasks` is called.
- This keeps the separation of concerns and avoids triggering `refreshTasks` within the board-loading logic unless explicitly needed.
  
This structure should now ensure that clicking the refresh button will reload both the board and the tasks for each column. Let me know if you need any more tweaks!

**Comment by tu2-atmanand** [comment]
I think earlier also, the second function was running if i call the first function only, because in console  i was able to see the following data : 

------ Inside the refreshBoardData function -----

Then the following line 5 to 6 times : 

------ Inside the refreshTasks function -----

**Comment by tu2-atmanand** [comment]
I think the first thing i can try for now is, create a function inside the services/ folder, to just read the tasks.json and refresh the currently opened board. This function you have to use after you do lot of functionalities.

The refreshBoardData function you can only use, if you have made any changes using boardConfigModal, after that the whole TaskBoard view should be get updated, how it happens now when you press the **Refresh Button**.

**Comment by tu2-atmanand** [comment]
Created a new function for the **Refresh Button**, now this button can re-render the whole Task Board View.
But this won't be suitable for updating the TaskBoard for simple stuffs like, after editing a task or adding a new task or after Scanning a Vault.

**Comment by tu2-atmanand** [comment]
- [x] **MayBe Best approach** : Create a function which will first removes everything from the tasks variable, and then it will run the setTask function.
  - This will somehow work if i first empty all the tasks from the board and then re-populate them. I can put some dealy also i guess.

**Comment by tu2-atmanand** [comment]
Alright, my understanding was correct somehow the above thing work, it was a simple solutoin, but i couldnt able to come up with earlier, now its working, and the task board not refresing after i edit any TaskItem Card has been resolved. Now the corresponding TaskItem card shows the updated content after edit modal.

**Comment by tu2-atmanand** [comment]
But there is still a problem of refreshing the Task Board after a new file has been scanned or after i have ran the Re-Scan Vault Modal. For this  i have come up with the following approach : 

See, for the functionalities like, whenever any new task has been detected, or other functoinalities, which i am not able to think as of now, but they will be (Rest i think has been taken care of and the board refreshes efficiently when any updations been done.), or functionalites like after i close the Re-Scan Vault modal, after i have scaned all the vault, this time only the tasks has to be updated and not the boards and everything(Which is happening right now when i call the `updateTasksAndRefreshBoard` at present, but that completely fine, since its happening very fast).
  - Firstly, remove all the code from the main.ts and create a new file called utils/realTimeScanning.ts and put it in this. Test this once you organize this code.
  - Now, whenever any new file has been sent for scanning, its tasks will be fetched and stored in the tasks.json. Now, i can do something like, send a message to the Column.tsx file, such that, it will run the `updateTasksAndRefreshBoard`, as soon as it receives this message. Similar thing i can do for any other file, like just import the Column.tsx from any file, and run/send message to this component and it will run that refresh function from there. So ask this to GPT and see if there is something in react, where i can send a message from any file to a component, becuase its not a class, so i cannot access any of its functoin, its a function component, so i think i will have to see if there is any feature like this only. Also, if obisidian provides a way like, earlier i was thinking of using localStorage, but that wont be possible, since its for reading any writing thing. The problem i want to solve is similar to kafka, like creating a channel and then sending request over this, where the receiver in this case the Column.tsx will receive it and will execute the function, and i want to run this function only from Column.tsx itself, since i think it only refreshes the specific column and also becuase it runs very effeciently. So, while writing a prompt to GPT, dont think like i want to call any func of Column.tsx, rather i want to send a mssg to Column.tsx once it receives this message it will execute the function.

**Comment by tu2-atmanand** [comment]
I have found a better method of EventEmmitter to send messages from any file within the project to refresh the board.
Now using the messages, i am re-rendering the board or column as required by different functionalities.
This issue has been resolved under the following PR : #32
